Core/include/Acts/Seeding/SeedFilter.ipp

```diff
@@ -113,12 +114,30 @@ void SeedFilter<external_spacepoint_t>::filterSeeds_1SpFixed(
                                           external_spacepoint_t>>>& i1,
                const std::pair<float, std::unique_ptr<const Acts::InternalSeed<
                                           external_spacepoint_t>>>& i2) {
-              return i1.first > i2.first;
+              if (i1.first > i2.first)
+                return true;
+              // This is for the case when the weights from different seeds are
+              // same It is added to make cpu & cuda results same
+              else if (i1.first == i2.first) {
```

Comment:
@robertlangenberg I added a condition, where the seed weights are same, to make cpu & cuda results same. It doesn't have any physical meaning inside though. It will be great if you can suggest a better idea about how deal with the seeds with same weight

Reply:
This is probably a VERY rare case, do you know how many cases like this occur? In case it's very very rare, I guess the computational overhead is miniscule. No need to take the square root though here, and generally use std::sqrt and std::pow instead of their C equivalents, please. For clarity's sake (and maybe a better branch prediction) write 

```c++
if (i1.first != i2.first){
  return i1.first != i2.first;
} else {
  // your code
}

Reply:
@robertlangenberg : I did not find the LIKELY/UNLIKLEY macros in ACTS that we use in Athena (coming from Gaudi). Maybe worth installing them?

Reply:
> This is probably a VERY rare case, do you know how many cases like this occur?
Without this condition, the seed matching rate becomes 100% -> ~99 %. So it should happen with a few percentage.

Reply:
@cgleggett I would like to point out two important caveats of LIKELY/UNLIKELY here.

First, under the hood, these macros map into compiler intrinsics. CUDA's nvcc may use different intrinsics on the GPU side than gcc/clang, or not have the intrinsics at all, and I'm pretty sure that Gaudi's implementation was never built for nvcc support. So if you want to use it on the GPU side, you probably won't be able to use the Gaudi LIKELY/UNLIKELY implementation as-is without modification.

Second, and perhaps more important. On all modern high-performance CPU architectures that I know of, the performance impact of a correctly predicted branch is identical no matter if the branch is taken or not taken. This essentially makes LIKELY/UNLIKELY obsolete as it doesn't matter if a compiler generates the most likely branch as taken or not taken. Contrast that to, say, the embedded MIPS architecture, where taken branches suffer a "branch delay slot" penalty and therefore compiler codegen hints are useful.

I'm not sure what the situation is on the GPU side. It could well be different because GPUs don't have branch predictors and speculation, they use their flavor of hyperthreading for branch latency hiding instead. But I would strongly advise you to roll out a quick CUDA microbenchmark where the control flow is convergent in the "taken" or "not taken" direction, and compare the performance, before resorting to manually annotating the code with possibly-obsolete compiler directives.

Reply:
@HadrienG2 
The above function does not run on the GPU so we dont have to benchmark branch performance with Cuda. It is the only part that was not parallelized in the Cuda Plugin.

Reply:
@beomki-yeo Thanks for the clarification! Then I guess there is even less motivation to use likely/unlikely macros here...

Reply:
@HadrienG2 : yes, I was only referring to CPU code here - you really wouldn't want to do anything like this on a GPU. Frankly though with all the security vulnerabilities like spectre due to branch mispredictions, I've lost track of which features Intel has disabled in the microarchitecture, so I'm not sure anymore if the compiler hints are useful or not.

Reply:
Yeah, when it gets to [extremes like these](https://www.phoronix.com/scan.php?page=news_item&px=LLVM-SESES-Merged), one has to wonder... :(

On the other hand, I feel pretty confident that we won't see an irreversible impact on the performance of predicted and taken branches anytime soon:

- Every tight loop contains an easily predictable and usually taken backwards branch in its assembly, so reducing the performance of that sort of branch means reducing the performance of almost every CPU benchmark in existence. I don't see Intel doing that in silicon or microcode, where there is no easy way to opt out.
- This leaves compiler-based mitigations. Here, cloud people can do arbitrarily silly things, but us developers have a way to counteract their evil deeds by passing in the right "I'm-not-a-cloud-instrastucture" compiler flag in the right place ;)

Reply:
> Without this condition, the seed matching rate becomes 100% -> ~99 %. So it should happen with a few percentage.

That is more than I expected, and this function is relatively hot. I think it should be fine if you change it as I suggested (removing the sqrt, which is quite expensive). You can also remove the middle SP from the calculation as it is the same for all seeds within a function call.

Reply:
@robertlangenberg  Thanks. Changes were made accordingly